exporter/loadbalancingexporter/resolver_k8s.go

```diff
@@ -0,0 +1,301 @@
+	ports []int32) (*k8sResolver, error) {
+
+	if len(service) == 0 {
+		return nil, errNoSvc
+	}
+	if len(ports) == 0 {
+		return nil, errNoSvcPorts
+	}
+
+	nAddr := strings.SplitN(service, ".", 2)
+	name, namespace := nAddr[0], "default"
+	if len(nAddr) > 1 {
+		namespace = nAddr[1]
+	}
+
+	epsSelector := fmt.Sprintf("metadata.name=%s", name)
+	epsListWatcher := &cache.ListWatch{
+		ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+			options.FieldSelector = epsSelector
+			return clt.CoreV1().Endpoints(namespace).List(context.TODO(), options)
+		},
+		WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+			options.FieldSelector = epsSelector
+			return clt.CoreV1().Endpoints(namespace).Watch(context.TODO(), options)
+		},
+	}
+
+	epsStore := &sync.Map{}
+	h := &handler{endpoints: epsStore, logger: logger}
+	r := &k8sResolver{
+		logger:         logger,
+		service:        name,
+		namespace:      namespace,
+		port:           ports,
+		once:           &sync.Once{},
+		endpointsStore: epsStore,
+		epsListWatcher: epsListWatcher,
+		handler:        h,
+		stopCh:         make(chan struct{}),
+	}
+	h.callback = r.resolve
+
+	return r, nil
+}
+
+func (r *k8sResolver) start(ctx context.Context) error {
+	if r.once == nil {
+		return fmt.Errorf("cannot Start() partial k8sResolver (nil *sync.Once)")
+	}
+	if r.handler == nil {
+		return fmt.Errorf("cannot Start() partial k8sResolver (nil *handler)")
+	}
+	var initErr error
+	r.once.Do(func() {
+		if r.epsListWatcher != nil {
+			r.logger.Debug("creating and starting endpoints informer")
+			epsInformer := cache.NewSharedInformer(r.epsListWatcher, &corev1.Endpoints{}, 0)
+			if _, err := epsInformer.AddEventHandler(r.handler); err != nil {
+				r.logger.Error("error adding event handler to endpoints informer", zap.Error(err))
+			}
+			go epsInformer.Run(r.stopCh)
+			if !cache.WaitForCacheSync(r.stopCh, epsInformer.HasSynced) {
+				r.logger.Error("error endpoints informer not sync")
+				initErr = errors.New("endpoints informer not sync")
+				return
+			}
+		}
+	})
+	if initErr != nil {
+		return initErr
+	}
+
+	r.logger.Debug("K8s service resolver started",
+		zap.String("service", r.service),
+		zap.String("namespace", r.namespace),
+		zap.Int32s("ports", r.port))
+	return nil
+}
+
+func (r *k8sResolver) shutdown(ctx context.Context) error {
+	r.changeCallbackLock.Lock()
+	r.onChangeCallbacks = nil
+	r.changeCallbackLock.Unlock()
+
+	close(r.stopCh)
+	r.shutdownWg.Wait()
+	return nil
+}
+func newInClusterClient() (kubernetes.Interface, error) {
+	cfg, err := config.GetConfig()
+	if err != nil {
+		return nil, err
+	}
+	return kubernetes.NewForConfig(cfg)
+}
+
+func (r *k8sResolver) resolve(ctx context.Context) ([]string, error) {
+	r.shutdownWg.Add(1)
+	defer r.shutdownWg.Done()
+
+	var backends []string
+	r.endpointsStore.Range(func(address, value any) bool {
+		addr := address.(string)
+		var backend string
+		ip, err := net.ResolveIPAddr("ip", addr)
```

Comment:
Why do you need their IPs? You can just use the hostnames, can't you? If the DNS is not updated with the service's IP, you'll fail here, but the consumer of this information (a gRPC client somewhere, most likely) will still have a chance to succeed, now or in a retry later on. So, I'd just go with the address, insted of the IP.

Reply:
i agree.